mlx/primitives.cpp

```diff
@@ -4348,6 +4348,107 @@ bool ScatterAxis::is_equivalent(const Primitive& other) const {
   return reduce_type_ == s_other.reduce_type_ && axis_ == s_other.axis_;
 }
 
+std::vector<array> MaskedScatter::vjp(
+    const std::vector<array>& primals,
+    const std::vector<array>& cotangents,
+    const std::vector<int>& argnums,
+    const std::vector<array>&) {
+  auto& s = stream();
+  const array& self = primals[0];
+  const array& mask = primals[1];
+  const array& src = primals[2];
+  const array mask_b = broadcast_to(mask, self.shape(), s);
+  const array& cotan = cotangents[0];
+
+  std::vector<array> vjps;
+  vjps.reserve(argnums.size());
+
+  for (int arg : argnums) {
+    if (arg == 0) {
+      vjps.push_back(where(mask_b, zeros_like(cotan, s), cotan, s));
+    } else if (arg == 2) {
+      const array mask_flat = flatten(mask_b, s);
+      const array cotan_flat = flatten(cotan, s);
+
+      const array mask_i = astype(mask_flat, int32, s);
+      const array prefix = cumsum(mask_i, 0, false, false, s);
+      const array idx_src = multiply(mask_i, prefix, s);
+      const array cotan_src =
+          multiply(cotan_flat, astype(mask_flat, cotan_flat.dtype(), s), s);
+
+      array gsrc_flat =
+          zeros({static_cast<int>(src.size())}, cotan_src.dtype(), s);
+      if (src.size() > 0) {
+        const int n = static_cast<int>(idx_src.size());
+        const array idx_vec = reshape(idx_src, {n}, s);
+        const array cotan_updates = reshape(cotan_src, {n, 1}, s);
+        gsrc_flat = scatter_add(gsrc_flat, idx_vec, cotan_updates, 0, s);
+      }
+
+      vjps.push_back(reshape(gsrc_flat, src.shape(), s));
+    } else {
+      throw std::invalid_argument(
+          "[masked_scatter] Cannot calculate VJP with respect to mask.");
+    }
+  }
+  return vjps;
+}
+
+std::vector<array> MaskedScatter::jvp(
+    const std::vector<array>& primals,
+    const std::vector<array>& tangents,
+    const std::vector<int>& argnums) {
+  auto& s = stream();
+  const array& self = primals[0];
+  const array& mask = primals[1];
+  const array mask_b = broadcast_to(mask, self.shape(), s);
+
+  array out = zeros_like(self, s);
+  for (int arg : argnums) {
+    if (arg == 0) {
+      out = where(mask_b, out, tangents[0], s);
+    } else if (arg == 2) {
+      out = masked_scatter(out, mask_b, tangents[1], s);
+      out = array(
+          self.shape(),
+          tangents[1].dtype(),
+          std::make_shared<MaskedScatter>(stream()),
+          {out, mask_b, tangents[1]});
```

Comment:
I'm confused by this. It looks like you are applying the same scatter twice? Am I missing something?